testing/file_test/file_test_base.cpp

```diff
@@ -533,14 +539,50 @@ static auto Main(int argc, char** argv) -> int {
     return EXIT_FAILURE;
   }
 
+  // Tests might try to read from stdin. Ensure those reads fail.
+  // TODO: This won't work on Windows.
+  int dev_null = open("/dev/null", O_RDONLY);
+  if (dev_null == -1) {
+    llvm::errs() << "Failed to open /dev/null.\n";
+    return EXIT_FAILURE;
+  }
+  if (dup2(dev_null, 0) != 0) {
+    llvm::errs() << "Failed to override stdin.\n";
+    return EXIT_FAILURE;
+  }
```

Comment:
Had you considered instead using `llvm::nulls` to replace the driver's stdin? Right now, it seems like there's a tradeoff where `bazel run` on a test which reads from stdin would wait on stdin: but such tests are rare, and doesn't `bazel run` hook up stdin? The trade-off of removing a cross-platform approach for incremental convenience with the two stdin tests seems difficult to me, but maybe we don't need to make that compromise.

Reply:
Using `llvm::nulls` doesn't help -- the driver is using `llvm::MemoryBuffer::getSTDIN`, which directly reads from file descriptor 0. (And we can't even mock that out or similar, because part of what we're trying to test is that we actually read from file descriptor 0 when given `-` as the input file.)

I've changed this to support Windows. I don't have any way to test it, but according to the documentation this should work. *shrug*

We could move the stdin tests off to a separate mechanism -- maybe an `sh_test` -- if you prefer. Or I can just undo this part of the change and leave `bazel run` on these testing targets broken. (`bazel run` seems to run the test in the background, so ^D at the terminal doesn't help, and bazel eats ^Cs so they don't help either; you need to find and `kill` the test wrapper script to unwedge a `bazel run` of the driver tests.)